# Error always happened

    const error = new Error('Whoop!');
    error.message; // Whoop!
    error.name; // Error
    error.stack; // "Error: Whoops! at <anonymous>:1:13"

错误通常是这样的: 

    Uncaught ErrorType: Error description

例如: 

```bash
    Uncaught TypeError: undefined is not a function
    Uncaught ReferenceError: Invalid left - hand side in assignment
    Unexpected token; 
    Uncaught SyntaxError: Unexpected token ILLEGAL
    Uncaught TypeError: Cannot read property' foo' of null, 
        Uncaught RangeError: Maximum call stack size exceeded
    XMLHttpRequest cannot load[http: //some/url/](http://some/url/). No 'Access-Control-Allow-Origin' header is present on the requested resource
```

### error table

| ERROR Type     | Error description                                                                  | Error Example                                                    |
|----------------|------------------------------------------------------------------------------------|------------------------------------------------------------------|
| InternalError  | 内部错误. 由 JavaScript 引擎抛出的错误                                                          |                                                                  |
| RangeError     | 范围错误. 发生在一个数值或参数超出其合法范围, 主要包括超出数组范围或者超出数字取值范围                                        | new Array(-1); (1234).toExponential(21); |
| ReferenceError | 引用错误. 通常是由于引用了一个不存在的值. | a.toString(); |
| SyntaxError    | 语法错误. | a ? a+1; |
| TypeError      | 类型错误. 通常是因为在执行特定的类型操作时, 变量的类型不符合要求. 例如变量中保存着意外类型, 访问不存在的方法等. | var a = new {}; var a = {a:1}; a.reverse(); // 对象并没有 reverser 方法 |
| URIError       | decodeURI() 或者 encodeURI() 传入非法参数时, 也包括 encodeURIComponent() 和 decodeURIComponent() | decodeURI('http://www.test.com&%'); encodeURIComponent('uD800'); |

### custom error

自定义 Error 类型

    class CustomError extends Error {
        constructor(message) {
            super(message);
            this.name = this.constructor.name;
            if (typeof Error.captureStackTrace === 'function') {
                // 在浏览器领域, 除了使用V8引擎的 Chrome, 
                // 其它浏览器中不存在 Error.captureStackTrace()这一接口, 
                // 所以在这里做一个条件判断. 
                Error.captureStackTrace(this, this.constructor);
                // 返回调用堆栈信息, 用于在 error 对象上添加合理的 stack 属性. 
            } else {
                this.stack = new Error(message).stack;
            }
        }
    }

### throw new Error

配合promise基本可以覆盖所有的请求情况了, 例如返回的结构信息是这样的:

```
response: {
    success: false,
    data: {},
    msg: 'error'
}
```

请求就可以这么写, as 在catch里可以统一处理请求失败或者数据不存在的情况. 

    request.then(res => {
        if (res && res.success && res.data) {
            // do something
        } else {
            throw new Error(res && res.msg)
        }
    }).catch(err => {
        showToast(error)
    })

